File: prolog/residual_gate.pl
:- module(residual_gate, [
    residual_disposition/2,
    evaluate_residual/2
]).

residual_disposition(absorbed, absorbed).
residual_disposition(provisional, provisional).
residual_disposition(commit_request, commit_request).

evaluate_residual(Residual, commit_request) :-
    Residual.get(structured, false) == true,
    ( Residual.get(recurrence_count, 0) > 1
    ; Residual.get(prediction_gain, 0.0) > 0.0
    ),
    !.
evaluate_residual(Residual, absorbed) :-
    Residual.get(disposition, provisional) == absorbed,
    !.
evaluate_residual(_, provisional).

File: python/omega_vision/core/residual_gate.py
"""SoW Appendix A.2 ``core/residual_gate.py`` / §7 — reduction and the salience gate.

Reads the description length the account leaves uncovered and turns it into the
three-way disposition ``absorbed | provisional | commit_request`` (SoW §7). The
gate decision is :class:`object_memory.memory.ResidualGate`; the analyzer that
separates unexplained structure from recognized transformations into
:class:`~object_memory.models.ResidualCandidate` objects is
:class:`object_memory.recognition.ResidualAnalyzer`.
"""

from object_memory.memory import ResidualGate
from object_memory.models import ResidualCandidate, ResidualDisposition
from object_memory.recognition import ResidualAnalyzer

__all__ = ["ResidualGate", "ResidualAnalyzer", "ResidualCandidate", "ResidualDisposition"]
